File: myPemoAnalysisFilterBank.py
# %%% myPemoAnalysisFilterBank.m %%%
import numpy as np
import scipy.signal as signal
from erbBW import erbBW
from gammatone_helper import Gfb_Analyzer_new, Gfb_Analyzer_process


# function [gfb_out_dec, analyzer,M] = myPemoAnalysisFilterBank(x,fs,M)
def myPemoAnalysisFilterBank(x, fs, Mmod=None):
    # MinCF = 20;
    MinCF = 20
    # MaxCF = fs/2;
    MaxCF = fs / 2.0
    # base_freq = 1000;
    base_freq = 1000
    # filters_per_ERB = 1.0;
    filters_per_ERB = 1.0

    # fsOrig = fs;
    fsOrig = fs
    # if fs/2 < 1.5*MaxCF,
    if fs / 2.0 < 1.5 * MaxCF:
        # x = resample(x, round(1.5*fs), fs);
        new_fs = int(round(1.5 * fs))
        x = signal.resample(x, int(round(len(x) * new_fs / fs)))
        # fs = round(1.5*fs);
        fs = new_fs

    # analyzer = Gfb_Analyzer_new(fs, MinCF, base_freq, MaxCF, filters_per_ERB);
    analyzer = Gfb_Analyzer_new(fs, MinCF, base_freq, MaxCF, filters_per_ERB)
    # analyzer.fsOrig = fsOrig;
    analyzer['fsOrig'] = fsOrig
    # [gfb_out, analyzer] = Gfb_Analyzer_process(analyzer, x(:).');
    gfb_out, analyzer = Gfb_Analyzer_process(analyzer, x)
    # Nb = size(gfb_out,1);
    Nb = gfb_out.shape[0]

    # if nargin<3 || isempty(M)
    if Mmod is None:
        # M = exp(-2*1i*pi/fs*analyzer.center_frequencies_hz(:)*(0:size(gfb_out,2)-1));
        time_steps = np.arange(gfb_out.shape[1])
        cfs = analyzer['center_frequencies_hz'][:, np.newaxis]
        Mmod = np.exp(-2j * np.pi / fs * cfs * time_steps)

    # gfb_out = gfb_out.*M;
    gfb_out = gfb_out * Mmod

    # bw = erbBW(analyzer.center_frequencies_hz);
    bw = erbBW(analyzer['center_frequencies_hz'])
    # alpha_dec = 2;
    alpha_dec = 2
    # Ndec = floor(fs./bw/alpha_dec);
    Ndec = np.floor(fs / (bw * alpha_dec)).astype(int)

    # analyzer.Ndec = Ndec;
    analyzer['Ndec'] = Ndec
    # analyzer.fs = fs;
    analyzer['fs'] = fs
    # analyzer.bw = bw;
    analyzer['bw'] = bw

    gfb_out_dec = []
    # for k=1:Nb
    for k in range(Nb):
        # gfb_out_dec{k} = resample(gfb_out(k,:),1,Ndec(k));
        decimated = signal.resample_poly(gfb_out[k, :], 1, Ndec[k])
        gfb_out_dec.append(decimated)

    return gfb_out_dec, analyzer, Mmod

File: references/peass_master_22c7fc4e/v1.0/myPemoAnalysisFilterBank.m
function [gfb_out_dec, analyzer,M] = myPemoAnalysisFilterBank(x,fs,M)
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Version 1.0
% Copyright 2010 Valentin Emiya (INRIA).
% This software is distributed under the terms of the GNU Public License
% version 3 (http://www.gnu.org/licenses/gpl.txt).
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

if ~isdeployed
    addpath(genpath('gammatone/'));
end

if nargin < 2,
    error('Not enough input arguments.');
end

% ----------------------------------------------------------------------
% settings
% ----------------------------------------------------------------------
MinCF = 20;				% (desired) minimum center frequency of the Gammatone filterbank in Hz
MaxCF = fs/2; % (desired) maximum center frequency of the Gammatone filterbank in Hz
base_freq = 1000;			% one of the gammatone filters will have this center frequency
filters_per_ERB = 3.0;		% density of gammatone filterbank

% upsampling of input signal to avoid aliasing in upper gammatone filters
fsOrig = fs;
if fs/2 < 1.5*MaxCF,
    x = resample(x, round(1.5*fs), fs);
    fs = round(1.5*fs);
end

% ----------------------------------------------------------------------
% actual signal processing
% ----------------------------------------------------------------------

% gammatone filterbank
analyzer = Gfb_Analyzer_new(fs, MinCF, base_freq, MaxCF, filters_per_ERB);
analyzer.fsOrig = fsOrig;
analyzer.fast = true;
[gfb_out, analyzer] = Gfb_Analyzer_process(analyzer, x(:).');
Nb = size(gfb_out,1);
if nargin<3 || isempty(M)
    M = exp(-2*1i*pi/fs*analyzer.center_frequencies_hz(:)*(0:size(gfb_out,2)-1));
end
gfb_out = gfb_out.*M;

% decimate
bw = erbBW(analyzer.center_frequencies_hz);
alpha_dec = 2;
gfb_out_dec = cell(Nb,1);
Ndec = floor(fs./bw/alpha_dec);
for k=1:Nb
    gfb_out_dec{k} = resample(gfb_out(k,:),1,Ndec(k));
end

if nargout>1
    analyzer.Ndec = Ndec;
    analyzer.fs = fs;
    analyzer.bw = bw;
end
return;


